File: lib-segment/src/main/java/websiteschema/mpsegment/hmm/Viterbi.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package websiteschema.mpsegment.hmm;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * reference: http://www.cs.umb.edu/~srevilak/viterbi/
 * @author ray
 */
public class Viterbi {

    NodeRepository stateBank = new NodeRepository();
    NodeRepository observeBank = new NodeRepository();
    Transition tran = new Transition();
    Pi pi = new Pi();
    Emission e = new Emission();
    int n = 2; // ngram


    public void setSortor(TrieNodeSortor sortor) {
        tran.setSortor(sortor);
    }

    public void setN(int n) {
        this.n = n;
    }

    public Emission getE() {
        return e;
    }

    public void setE(Emission e) {
        this.e = e;
    }

    public NodeRepository getObserveBank() {
        return observeBank;
    }

    public void setObserveBank(NodeRepository observeBank) {
        this.observeBank = observeBank;
    }

    public Pi getPi() {
        return pi;
    }

    public void setPi(Pi pi) {
        this.pi = pi;
    }

    public NodeRepository getStateBank() {
        return stateBank;
    }

    public void setStateBank(NodeRepository stateBank) {
        this.stateBank = stateBank;
    }

    public Transition getTran() {
        return tran;
    }

    public void setTran(Transition tran) {
        this.tran = tran;
    }

    public static int[] getStatePath(int[][] states, int[][] psai, int end, int depth, int pos) {
        int maxDepth = end + 1 > depth ? depth : end + 1;
        int[] ret = new int[maxDepth];

        for (int i = 0; i < maxDepth; i++) {
            int state = states[end - i][pos];
            pos = psai[end - i][pos];
            ret[ret.length - i - 1] = state;
        }

        return ret;
    }

    private HmmResult calculateHmmResult(List<String> listObserve) throws ObserveListException {
        HmmResult ret = new HmmResult();

        if (listObserve.isEmpty()) {
            throw new ObserveListException("observe list is empty.");
        }

        String o = listObserve.get(0);
        Node o1 = observeBank.get(o);
        if (o1 == null) {
            o1 = new Node(o);
            observeBank.add(o1);
        }

        Set<Integer> relatedStates = e.getStateProbByObserve(o1.getIndex());
        if (null == relatedStates || relatedStates.isEmpty()) {
            throw new ObserveListException("UNKNOWN observe object " + o + ".");
        }
        ret.states = new int[listObserve.size()][];
        ret.delta = new double[listObserve.size()][];
        ret.psai = new int[listObserve.size()][];
        ret.states[0] = new int[relatedStates.size()];
        ret.delta[0] = new double[relatedStates.size()];
        ret.psai[0] = new int[relatedStates.size()];
        int index = 0;
        for (Integer s : relatedStates) {
            ret.states[0][index] = s;
            ret.delta[0][index] = Math.log(pi.getPi(s)) + Math.log(e.getProb(s, o1.getIndex()));
            ret.psai[0][index] = 0;
            index++;
        }

        //
        for (int p = 1; p < listObserve.size(); p++) {
            o = listObserve.get(p);
            Node oi = observeBank.get(o);
            if (oi == null) {
                oi = new Node(o);
                observeBank.add(oi);
            }

            Set<Integer> stateSet = e.getStateProbByObserve(oi.getIndex());
            if (stateSet.isEmpty()) {
                throw new ObserveListException("UNKNOWN observe object " + o + ".");
            }
            ret.states[p] = new int[stateSet.size()];
            ret.delta[p] = new double[stateSet.size()];
            ret.psai[p] = new int[stateSet.size()];
            int i = 0;
            for (int state : stateSet) {
                ret.states[p][i] = state;
                double maxDelta = Double.NEGATIVE_INFINITY;
                double maxPsai = Double.NEGATIVE_INFINITY;
                int ls = 0;
                for (int j = 0; j < ret.states[p - 1].length; j++) {
                    int[] statePath = getStatePath(ret.states, ret.psai, p - 1, n - 1, j);
                    double b = Math.log(e.getProb(state, oi.getIndex()));
                    double Aij = Math.log(tran.getCoProb(statePath, state));
                    double psai_j = ret.delta[p - 1][j] + Aij;
                    double delta_j = psai_j + b;
                    if (delta_j > maxDelta) {
                        maxDelta = delta_j;
                    }

                    if (psai_j > maxPsai) {
                        maxPsai = psai_j;
                        ls = j;
                    }
                }

                ret.delta[p][i] = maxDelta;
                ret.psai[p][i] = ls;

                i++;
            }
        }

        return ret;
    }

    public List<Node> calculateWithLog(List<String> listObserve) throws ObserveListException {
        HmmResult ret = calculateHmmResult(listObserve);
        double maxProb = Double.NEGATIVE_INFINITY;
        int pos = 0;
        for (int j = 0; j < ret.delta[listObserve.size() - 1].length; j++) {
            double p = ret.delta[listObserve.size() - 1][j];
            if (p > maxProb) {
                maxProb = p;
                pos = j;
            }
        }

        int[] statePath = getStatePath(ret.states, ret.psai, listObserve.size() - 1, listObserve.size(), pos);
        List<Node> path = new ArrayList<Node>();
        for (int state : statePath) {
            path.add(stateBank.get(state));
        }

        return path;
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/hmm/Viterbi.scala
package websiteschema.mpsegment.hmm

import java.util


/**
 * reference: http://www.cs.umb.edu/~srevilak/viterbi/
 */

trait Viterbi {

  def getConditionProb(statePath: Array[Int], state: Int): Double

  def getProb(state: Int, observe: Int): Double

  def getPi(state: Int): Double

  def getStatesBy(observe: Int): java.util.Collection[Int]

  def calculateWithLog(listObserve: Seq[String]): Seq[Node]

  def setObserveBank(observeBank: NodeRepository)
  def setStateBank(stateBank: NodeRepository)
  def setTran(tran: ITransition)
  def setE(e: Emission)
  def setPi(pi: Pi)
}

class ViterbiImpl extends Viterbi {

  var stateBank = new NodeRepository()
  var observeBank = new NodeRepository()
  var tran: ITransition = Transition()
  var pi = Pi()
  var e = Emission()
  var n = 2

  def setN(n: Int) {
    this.n = n
  }

  def getE(): Emission = {
    return e
  }

  def setE(e: Emission) {
    this.e = e
  }

  def getObserveBank(): NodeRepository = {
    return observeBank
  }

  def setObserveBank(observeBank: NodeRepository) {
    this.observeBank = observeBank
  }

  def getPi(): Pi = {
    return pi
  }

  def setPi(pi: Pi) {
    this.pi = pi
  }

  def getStateBank(): NodeRepository = {
    return stateBank
  }

  def setStateBank(stateBank: NodeRepository) {
    this.stateBank = stateBank
  }

  def setTran(tran: ITransition) {
    this.tran = tran
  }

  def getStatePath(states: Array[Array[Int]], psai: Array[Array[Int]], end: Int, depth: Int, position: Int): Array[Int] = {
    val maxDepth = if (end + 1 > depth) depth else end + 1
    val ret = new Array[Int](maxDepth)
    var pos = position
    for (i <- 0 until maxDepth) {
      val state = states(end - i)(pos)
      pos = psai(end - i)(pos)
      ret(ret.length - i - 1) = state
    }
    ret
  }

  private def getObserveIndex(observe: String): Int = {
    val node = observeBank.get(observe)
    if (node != null) {
      node.getIndex()
    } else {
      val newNode = Node(observe)
      observeBank.add(newNode)
      newNode.getIndex()
    }
  }


  private def calculateResult(listObserve: Seq[String]): ViterbiResult = {
    val ret = new ViterbiResult()

    if (listObserve.isEmpty) {
      throw new ObserveListException("observe list is empty.")
    }

    val o1 = getObserveIndex(listObserve(0))
    val relatedStates = getStatesBy(o1)

    val length = listObserve.size
    ret.states = new Array[Array[Int]](length)
    ret.delta = new Array[Array[Double]](length)
    ret.psai = new Array[Array[Int]](length)

    val relatedStatesCount = relatedStates.size
    initResultInPosition(ret, 0, relatedStatesCount)

    var index = 0
    val iterRlatedStates = relatedStates.iterator()
    while (iterRlatedStates.hasNext) {
      val s = iterRlatedStates.next()
      ret.states(0)(index) = s
      ret.delta(0)(index) = Math.log(getPi(s)) + Math.log(getProb(s, o1))
      ret.psai(0)(index) = 0
      index += 1
    }

    for (p <- 1 until listObserve.size) {
      val o = listObserve(p)
      var oi = observeBank.get(o)
      if (oi == null) {
        oi = Node(o)
        observeBank.add(oi)
      }

      val stateSet = getStatesBy(oi.getIndex())
      initResultInPosition(ret, p, stateSet.size)
      var i = 0
      val iterStateSet = stateSet.iterator()
      while (iterStateSet.hasNext) {
        val state = iterStateSet.next()
        ret.states(p)(i) = state
        var maxDelta = Double.NegativeInfinity
        var maxPsai = Double.NegativeInfinity
        var ls = 0
        var j = 0
        while (j < ret.states(p - 1).length) {
          val statePath = getStatePath(ret.states, ret.psai, p - 1, n - 1, j)
          val b = Math.log(getProb(state, oi.getIndex()))
          val Aij = Math.log(getConditionProb(statePath, state))
          val psai_j = ret.delta(p - 1)(j) + Aij
          val delta_j = psai_j + b
          if (delta_j > maxDelta) {
            maxDelta = delta_j
          }

          if (psai_j > maxPsai) {
            maxPsai = psai_j
            ls = j
          }
          j += 1
        }

        ret.delta(p)(i) = maxDelta
        ret.psai(p)(i) = ls

        i += 1
      }
    }

    return ret
  }

  def initResultInPosition(ret: ViterbiResult, position: Int, relatedStatesCount: Int) {
    ret.states(position) = new Array[Int](relatedStatesCount)
    ret.delta(position) = new Array[Double](relatedStatesCount)
    ret.psai(position) = new Array[Int](relatedStatesCount)
  }

  override def getConditionProb(statePath: Array[Int], state: Int) = tran.getConditionProb(statePath, state)

  override def getProb(state: Int, observe: Int) = e.getProb(state, observe)

  override def getPi(state: Int) = pi.getPi(state)

  override def getStatesBy(observe: Int) = {
    val states = e.getStatesBy(observe)
    if (null == states || states.isEmpty) {
      throw new ObserveListException("UNKNOWN observe object " + observe + ".")
    }
    states
  }

  override def calculateWithLog(listObserve: Seq[String]): Seq[Node] = {
    if (listObserve.isEmpty) {
      List[Node]()
    } else {
      val ret = calculateResult(listObserve)
      var maxProb = Double.NegativeInfinity
      var pos = 0
      for (j <- 0 until ret.delta(listObserve.size - 1).length) {
        val p = ret.delta(listObserve.size - 1)(j)
        if (p > maxProb) {
          maxProb = p
          pos = j
        }
      }

      val statePath = getStatePath(ret.states, ret.psai, listObserve.size - 1, listObserve.size, pos)
      statePath.map(stateBank.get(_))
    }
  }
}
